File: android/app/src/main/java/com/voxa/app/viewmodel/DMViewModel.kt
package com.voxa.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.voxa.app.data.api.ApiClient
import com.voxa.app.data.api.SocketClient
import com.voxa.app.data.model.DMChannel
import com.voxa.app.data.model.DMMessage
import com.voxa.app.data.model.User
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class DMViewModel : ViewModel() {

    private val _channels = MutableStateFlow<List<DMChannel>>(emptyList())
    val channels: StateFlow<List<DMChannel>> = _channels

    private val _messages = MutableStateFlow<Map<String, List<DMMessage>>>(emptyMap())
    val messages: StateFlow<Map<String, List<DMMessage>>> = _messages

    private val _isLoadingChannels = MutableStateFlow(false)
    val isLoadingChannels: StateFlow<Boolean> = _isLoadingChannels

    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    private var pollJob: Job? = null
    private val joinedRooms = mutableSetOf<String>()

    fun load() {
        viewModelScope.launch {
            _isLoadingChannels.value = true
            ApiClient.dmChannels()
                .onSuccess { chans ->
                    _channels.value = chans
                    chans.forEach { ch ->
                        SocketClient.joinDM(ch.id)
                        joinedRooms.add(ch.id)
                    }
                }
                .onFailure { _error.value = it.message }
            _isLoadingChannels.value = false
            setupSocketCallbacks()
            startPolling()
        }
    }

    fun loadMessages(dmId: String) {
        if (_messages.value[dmId] != null) return
        viewModelScope.launch {
            ApiClient.dmMessages(dmId)
                .onSuccess { msgs ->
                    _messages.value = _messages.value + (dmId to msgs)
                }
        }
    }

    fun refreshMessages(dmId: String) {
        viewModelScope.launch {
            ApiClient.dmMessages(dmId)
                .onSuccess { msgs ->
                    _messages.value = _messages.value + (dmId to msgs)
                }
        }
    }

    fun openDM(username: String, onResult: (DMChannel?) -> Unit) {
        viewModelScope.launch {
            ApiClient.openDm(username)
                .onSuccess { ch ->
                    if (_channels.value.none { it.id == ch.id }) {
                        _channels.value = listOf(ch) + _channels.value
                        SocketClient.joinDM(ch.id)
                        joinedRooms.add(ch.id)
                    }
                    onResult(ch)
                }
                .onFailure {
                    _error.value = it.message
                    onResult(null)
                }
        }
    }

    fun send(content: String, channel: DMChannel, author: User) {
        if (content.isBlank()) return
        viewModelScope.launch {
            _isSending.value = true
            val optimistic = DMMessage(
                id = "opt_${System.currentTimeMillis()}",
                dmChannelId = channel.id,
                authorId = author.id,
                author = author.username,
                displayName = author.displayName,
                avatarColor = author.avatarColor,
                content = content,
                edited = false,
                timestamp = java.time.Instant.now().toString(),
            )
            appendMessage(channel.id, optimistic)

            ApiClient.sendDmMessage(channel.id, content)
                .onSuccess { real ->
                    val current = _messages.value[channel.id]?.toMutableList() ?: mutableListOf()
                    val idx = current.indexOfFirst { it.id == optimistic.id }
                    if (idx >= 0) current[idx] = real else current.add(real)
                    _messages.value = _messages.value + (channel.id to current)
                    bumpLastMessage(channel.id, real)
                }
            _isSending.value = false
        }
    }

    fun edit(message: DMMessage, newContent: String, dmId: String) {
        viewModelScope.launch {
            _messages.value = _messages.value + (dmId to
                (_messages.value[dmId] ?: emptyList()).map {
                    if (it.id == message.id) it.copy(content = newContent, edited = true) else it
                })
            ApiClient.editDmMessage(dmId, message.id, newContent)
                .onSuccess { updated ->
                    _messages.value = _messages.value + (dmId to
                        (_messages.value[dmId] ?: emptyList()).map {
                            if (it.id == message.id) updated else it
                        })
                }
        }
    }

    fun delete(message: DMMessage, dmId: String) {
        _messages.value = _messages.value + (dmId to
            (_messages.value[dmId] ?: emptyList()).filter { it.id != message.id })
        viewModelScope.launch { ApiClient.deleteDmMessage(dmId, message.id) }
    }

    fun markRead(dmId: String) {
        viewModelScope.launch { ApiClient.markDmRead(dmId) }
    }

    fun clearError() { _error.value = null }

    fun disconnect() {
        joinedRooms.forEach { SocketClient.leaveDM(it) }
        joinedRooms.clear()
        pollJob?.cancel()
        pollJob = null
        SocketClient.onDMMessage = null
        SocketClient.onDMMessageEdit = null
        SocketClient.onDMMessageDelete = null
    }

    private fun setupSocketCallbacks() {
        SocketClient.onDMMessage = { msg ->
            val current = _messages.value[msg.dmChannelId]?.toMutableList() ?: mutableListOf()
            val optIdx = current.indexOfFirst {
                it.id.startsWith("opt_") && it.content == msg.content && it.authorId == msg.authorId
            }
            if (optIdx >= 0) current[optIdx] = msg
            else if (current.none { it.id == msg.id }) current.add(msg)
            _messages.value = _messages.value + (msg.dmChannelId to current)
            bumpLastMessage(msg.dmChannelId, msg)
        }
        SocketClient.onDMMessageEdit = { msg ->
            _messages.value = _messages.value + (msg.dmChannelId to
                (_messages.value[msg.dmChannelId] ?: emptyList()).map {
                    if (it.id == msg.id) msg else it
                })
        }
        SocketClient.onDMMessageDelete = { id, dmChannelId ->
            _messages.value = _messages.value + (dmChannelId to
                (_messages.value[dmChannelId] ?: emptyList()).filter { it.id != id })
        }
    }

    private fun appendMessage(dmId: String, msg: DMMessage) {
        val current = _messages.value[dmId]?.toMutableList() ?: mutableListOf()
        current.add(msg)
        _messages.value = _messages.value + (dmId to current)
    }

    private fun bumpLastMessage(dmId: String, msg: DMMessage) {
        _channels.value = _channels.value.map {
            if (it.id == dmId) it.copy(lastMessage = msg) else it
        }
    }

    private fun startPolling() {
        pollJob?.cancel()
        pollJob = viewModelScope.launch {
            while (isActive) {
                delay(15_000)
                // Refresh messages for channels that are loaded
                _messages.value.keys.forEach { dmId ->
                    ApiClient.dmMessages(dmId).onSuccess { fresh ->
                        _messages.value = _messages.value + (dmId to fresh)
                    }
                }
            }
        }
    }
}

File: ios/Voxa/ViewModels/DMViewModel.swift
import Foundation
import SwiftUI

@MainActor
class DMViewModel: ObservableObject {
    @Published var channels: [DMChannel] = []
    @Published var selectedChannel: DMChannel?
    @Published var messages: [String: [DMMessage]] = [:]  // dmId → messages
    @Published var isLoadingChannels = false
    @Published var isSending = false
    @Published var error: String?

    private let socket = SocketClient.shared
    private var pollTask: Task<Void, Never>?
    private var joinedRooms: Set<String> = []

    func load() async {
        isLoadingChannels = true
        do {
            channels = try await APIClient.shared.dmChannels()
            for ch in channels {
                socket.joinDM(ch.id)
                joinedRooms.insert(ch.id)
            }
        } catch {
            self.error = error.localizedDescription
        }
        isLoadingChannels = false

        setupSocketCallbacks()
        startPolling()
    }

    func selectChannel(_ channel: DMChannel) {
        selectedChannel = channel
        if messages[channel.id] == nil {
            Task { await loadMessages(dmId: channel.id) }
        }
        socket.joinDM(channel.id)
        joinedRooms.insert(channel.id)
        Task { try? await APIClient.shared.markDmRead(dmId: channel.id) }
    }

    func loadMessages(dmId: String) async {
        do {
            let msgs = try await APIClient.shared.dmMessages(dmId: dmId)
            messages[dmId] = msgs
        } catch {}
    }

    func openDM(username: String) async -> DMChannel? {
        do {
            let ch = try await APIClient.shared.openDm(username: username)
            if !channels.contains(where: { $0.id == ch.id }) {
                channels.insert(ch, at: 0)
                socket.joinDM(ch.id)
                joinedRooms.insert(ch.id)
            }
            return ch
        } catch {
            self.error = error.localizedDescription
            return nil
        }
    }

    func send(content: String, in channel: DMChannel, author: User) async {
        guard !content.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        isSending = true

        // Optimistic
        let optimistic = DMMessage(
            id: "opt_\(UUID().uuidString)",
            dmChannelId: channel.id,
            authorId: author.id,
            author: author.username,
            displayName: author.displayName,
            avatarUrl: author.avatarUrl,
            avatarColor: author.avatarColor,
            content: content,
            edited: false,
            editedAt: nil,
            timestamp: Date()
        )
        var msgs = messages[channel.id] ?? []
        msgs.append(optimistic)
        messages[channel.id] = msgs

        do {
            let real = try await APIClient.shared.sendDmMessage(dmId: channel.id, content: content)
            if var current = messages[channel.id],
               let idx = current.firstIndex(where: { $0.id == optimistic.id }) {
                current[idx] = real
                messages[channel.id] = current
            }
            bumpLastMessage(real, in: channel.id)
        } catch {}
        isSending = false
    }

    func edit(message: DMMessage, newContent: String, in channel: DMChannel) async {
        guard var msgs = messages[channel.id],
              let idx = msgs.firstIndex(where: { $0.id == message.id }) else { return }
        msgs[idx].content = newContent
        msgs[idx].edited = true
        messages[channel.id] = msgs

        do {
            let updated = try await APIClient.shared.editDmMessage(
                dmId: channel.id, msgId: message.id, content: newContent)
            if var current = messages[channel.id],
               let i = current.firstIndex(where: { $0.id == message.id }) {
                current[i] = updated
                messages[channel.id] = current
            }
        } catch {}
    }

    func delete(message: DMMessage, in channel: DMChannel) async {
        messages[channel.id]?.removeAll { $0.id == message.id }
        try? await APIClient.shared.deleteDmMessage(dmId: channel.id, msgId: message.id)
    }

    func disconnect() {
        for id in joinedRooms { socket.leaveDM(id) }
        joinedRooms = []
        pollTask?.cancel()
        pollTask = nil
        socket.onDMMessage = nil
        socket.onDMMessageEdit = nil
        socket.onDMMessageDelete = nil
    }

    // MARK: - Private

    private func setupSocketCallbacks() {
        socket.onDMMessage = { [weak self] msg in
            guard let self else { return }
            var msgs = self.messages[msg.dmChannelId] ?? []
            // Replace optimistic or append
            if let optIdx = msgs.firstIndex(where: {
                $0.id.hasPrefix("opt_") && $0.content == msg.content && $0.authorId == msg.authorId
            }) {
                msgs[optIdx] = msg
            } else if !msgs.contains(where: { $0.id == msg.id }) {
                msgs.append(msg)
            }
            self.messages[msg.dmChannelId] = msgs
            self.bumpLastMessage(msg, in: msg.dmChannelId)
        }

        socket.onDMMessageEdit = { [weak self] msg in
            guard let self else { return }
            if var msgs = self.messages[msg.dmChannelId],
               let idx = msgs.firstIndex(where: { $0.id == msg.id }) {
                msgs[idx] = msg
                self.messages[msg.dmChannelId] = msgs
            }
        }

        socket.onDMMessageDelete = { [weak self] id, dmChannelId in
            guard let self else { return }
            self.messages[dmChannelId]?.removeAll { $0.id == id }
        }
    }

    private func bumpLastMessage(_ msg: DMMessage, in dmId: String) {
        if let idx = channels.firstIndex(where: { $0.id == dmId }) {
            channels[idx].lastMessage = msg
        }
    }

    private func startPolling() {
        pollTask?.cancel()
        pollTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 15_000_000_000) // 15s fallback
                guard !Task.isCancelled else { break }
                if let sel = selectedChannel {
                    if let fresh = try? await APIClient.shared.dmMessages(dmId: sel.id) {
                        mergeMessages(fresh, dmId: sel.id)
                    }
                }
            }
        }
    }

    private func mergeMessages(_ fresh: [DMMessage], dmId: String) {
        var merged = messages[dmId] ?? []
        for msg in fresh {
            if let idx = merged.firstIndex(where: { $0.id == msg.id }) {
                merged[idx] = msg
            } else if !merged.contains(where: { $0.id == msg.id }) {
                merged.append(msg)
            }
        }
        merged.removeAll { msg in
            msg.id.hasPrefix("opt_") &&
            fresh.contains(where: { $0.content == msg.content && $0.authorId == msg.authorId })
        }
        messages[dmId] = merged.sorted { $0.timestamp < $1.timestamp }
    }
}
